src/grav/test.py

```diff
@@ -0,0 +1,115 @@
+"""
+This file contains meat of the Gravitational QA coding challenge.
+
+The test() function performs the test as requested in the challenge.
+
+The setup() function does all the provisioning necessary to go from:
+  "package is installed" -> "start/stop/logs a wellknown container"
+
+The teardown() function attempts to unwind any dangling resources
+from setup() and test(). Ideally leaving a clean test enviromnent
+if it runs to completion.  This is important for idempotency.
+"""
+import re
+from pkg_resources import resource_filename
+
+from grav.client import demux_logs, DockerClient
+
+
+def setup():
+    # If this were real, I'd pass these magic strings stattered through
+    # file as params and have them loaded from a config.
+    image_tar = resource_filename(__name__, "wellknown.tar")
+
+    # It'd be better to nest client, image, container setup/teardown
+    # so that in the case client & image succeed but container fails
+    # we only run the teardown for image & client, and skip container
+    # teardown since it was never set up.
+    #
+    # That is more effort than this sample merits imo, so setup/teardown
+    # are all in one. -- wdella 2019-10
+    client = DockerClient("/var/run/docker.sock")
+
+    # load image from tar, get image id
+    with open(image_tar, 'rb') as fh:
+        data = fh.read()
+
+    headers = {'Content-Type': 'application/tar', 'Content-Length': len(data)}
+    resp = client.post("/images/load?fromSrc=-", body=data, headers=headers)
+    assert resp.status in (200, 201)
+    json = resp.json()
+    match = re.match(r"Loaded image ID: (sha256:[\da-f]{64})", json['stream'])
+    image_id = match.group(1)
+
+    # create container, get container id
+    data = {"Image": image_id}
+    headers = {"Content-Type": "application/json"}
+    resp = client.post("/containers/create", body=data, headers=headers)
+    assert resp.status in (200, 201)
+    json = resp.json()
+    container_id = json['Id']
+
+    return client, image_id, container_id
+
+
+def test(client, container_id):
+    # start
+    start_path = "/containers/%s/start" % container_id
```

Comment:
Another idea similar to the one above, is you could make a thin wrapper over your low-level client which would provide `start()`, `stop()` and `logs()` methods to keep tests focused on testing the logic rather than building URLs, configuring headers, etc.

Reply:
I considered having a higher level wrapper -- especially with all this repeated logic --, but it didn't make the priority cut for this PR.  I propose filing an issue for this which can be triaged and implemented at the appropriate priority level. (aka, this is a great idea, but I'm not sure if it in scope).

Reply:
https://github.com/javins/gravitational-audition/issues/6 filed to track this.